core/src/repair_service.rs

```diff
@@ -262,29 +268,51 @@ impl RepairService {
                 )
             };
 
-            let mut send_repairs_elapsed = Measure::start("send_repairs_elapsed");
+            let mut build_repairs_batch_elapsed = Measure::start("build_repairs_batch_elapsed");
             let mut outstanding_requests = outstanding_requests.write().unwrap();
-            repairs.into_iter().for_each(|repair_request| {
-                if let Ok((to, req)) = serve_repair.repair_request(
-                    &repair_info.cluster_slots,
-                    repair_request,
-                    &mut peers_cache,
-                    &mut repair_stats,
-                    &repair_info.repair_validators,
-                    &mut outstanding_requests,
-                ) {
-                    repair_socket.send_to(&req, to).unwrap_or_else(|e| {
-                        info!("{} repair req send_to({}) error {:?}", id, to, e);
-                        0
-                    });
+            let batch: Vec<(Vec<u8>, SocketAddr)> = repairs
+                .iter()
+                .filter_map(|repair_request| {
+                    if let Ok((to, req)) = serve_repair.repair_request(
+                        &repair_info.cluster_slots,
+                        *repair_request,
+                        &mut peers_cache,
+                        &mut repair_stats,
+                        &repair_info.repair_validators,
+                        &mut outstanding_requests,
+                    ) {
+                        Some((req, to))
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            drop(outstanding_requests);
```

Comment:
I think you can utilize `?` operator here. Also giving `outstanding_requests` a scope will avoid manual `drop`. So something like:
```rust
let batch: Vec<_> = {
  let mut outstanding_requests = outstanding_requests.write().unwrap();
  repairs
    .into_iter()
    .filter_map(|repair_request| {
        let (to, req) = serve_repair.repair_request(...).ok()?;
        Some((req, to))
    })
    .collect()
};
```

Reply:
That's more succinct, changed.